In [1]:
# =============================================================================
# STEP 1: LOAD PROCESSED DATASET
# =============================================================================

import pandas as pd
import numpy as np


# Load cleaned dataset
df = pd.read_csv(
    "../../../data/processed/heartdisease_cleaned.csv"
)


print("=" * 70)
print("PROCESSED DATASET")
print("=" * 70)

print(f"Rows    : {df.shape[0]}")
print(f"Columns : {df.shape[1]}")


display(df.head())

PROCESSED DATASET
Rows    : 303
Columns : 14


,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63.0,1.0,1.0,145.0,233.0,1.0,2.0,150.0,0.0,2.3,3.0,0.0,6.0,0
1,67.0,1.0,4.0,160.0,286.0,0.0,2.0,108.0,1.0,1.5,2.0,3.0,3.0,1
2,67.0,1.0,4.0,120.0,229.0,0.0,2.0,129.0,1.0,2.6,2.0,2.0,7.0,1
3,37.0,1.0,3.0,130.0,250.0,0.0,0.0,187.0,0.0,3.5,3.0,0.0,3.0,0
4,41.0,0.0,2.0,130.0,204.0,0.0,2.0,172.0,0.0,1.4,1.0,0.0,3.0,0


In [2]:
# =============================================================================
# STEP 2: FEATURE AND TARGET SEPARATION
# =============================================================================

target_column = "target"


X = df.drop(
    target_column,
    axis=1
)


y = df[target_column]


print("=" * 70)
print("FEATURE MATRIX (X)")
print("=" * 70)

print(f"Rows    : {X.shape[0]}")
print(f"Columns : {X.shape[1]}")


print("\n" + "=" * 70)
print("TARGET VARIABLE (y)")
print("=" * 70)

print(f"Rows : {y.shape[0]}")


display(y.head())

FEATURE MATRIX (X)
Rows    : 303
Columns : 13

TARGET VARIABLE (y)
Rows : 303


0    0
1    1
2    1
3    0
4    0
Name: target, dtype: int64

In [3]:
# =============================================================================
# STEP 3: VERIFY TARGET DISTRIBUTION
# =============================================================================

print("=" * 70)
print("TARGET DISTRIBUTION")
print("=" * 70)


display(
    y.value_counts()
)


print("\nClass Percentage")


display(
    (y.value_counts(normalize=True) * 100)
    .round(2)
)

TARGET DISTRIBUTION


target
0    164
1    139
Name: count, dtype: int64


Class Percentage


target
0    54.13
1    45.87
Name: proportion, dtype: float64

In [4]:
# =============================================================================
# STEP 4: FEATURE TYPE CHECK
# =============================================================================

print("=" * 70)
print("FEATURE DATA TYPES")
print("=" * 70)


display(
    X.dtypes.to_frame(
        name="Data Type"
    )
)

FEATURE DATA TYPES


,Data Type
age,float64
sex,float64
cp,float64
trestbps,float64
chol,float64
fbs,float64
restecg,float64
thalach,float64
exang,float64
oldpeak,float64


In [5]:
# =============================================================================
# STEP 5: MISSING VALUE CHECK
# =============================================================================

print("=" * 70)
print("MISSING VALUES")
print("=" * 70)


missing = pd.DataFrame({
    "Missing Count": X.isnull().sum(),
    "Missing Percentage (%)": (
        X.isnull().mean() * 100
    ).round(2)
})


display(
    missing[missing["Missing Count"] > 0]
)

MISSING VALUES


,Missing Count,Missing Percentage (%)
ca,4,1.32
thal,2,0.66


In [6]:
# =============================================================================
# STEP 6: MISSING VALUE IMPUTATION
# =============================================================================

from sklearn.impute import SimpleImputer


imputer = SimpleImputer(
    strategy="median"
)


X_imputed = pd.DataFrame(
    imputer.fit_transform(X),
    columns=X.columns,
    index=X.index
)


print("=" * 70)
print("MISSING VALUE IMPUTATION")
print("=" * 70)

print("Imputation completed successfully")


print("\nRemaining Missing Values:")
print(X_imputed.isnull().sum().sum())

MISSING VALUE IMPUTATION
Imputation completed successfully

Remaining Missing Values:
0


In [7]:
# =============================================================================
# STEP 7: TRAIN-TEST SPLIT
# =============================================================================

from sklearn.model_selection import train_test_split


X_train, X_test, y_train, y_test = train_test_split(
    X_imputed,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)


print("=" * 70)
print("TRAIN-TEST SPLIT")
print("=" * 70)


print("Training Data")
print("-" * 70)

print(f"X_train shape : {X_train.shape}")
print(f"y_train shape : {y_train.shape}")


print("\nTesting Data")
print("-" * 70)

print(f"X_test shape  : {X_test.shape}")
print(f"y_test shape  : {y_test.shape}")

TRAIN-TEST SPLIT
Training Data
----------------------------------------------------------------------
X_train shape : (242, 13)
y_train shape : (242,)

Testing Data
----------------------------------------------------------------------
X_test shape  : (61, 13)
y_test shape  : (61,)


In [8]:
# =============================================================================
# STEP 8: CLASS DISTRIBUTION AFTER SPLIT
# =============================================================================

print("=" * 70)
print("TRAINING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_train.value_counts(normalize=True) * 100)
    .round(2)
)


print("\n" + "=" * 70)
print("TESTING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_test.value_counts(normalize=True) * 100)
    .round(2)
)

TRAINING TARGET DISTRIBUTION


target
0    54.13
1    45.87
Name: proportion, dtype: float64


TESTING TARGET DISTRIBUTION


target
0    54.1
1    45.9
Name: proportion, dtype: float64

In [9]:
# =============================================================================
# STEP 9: FEATURE SCALING
# =============================================================================

from sklearn.preprocessing import StandardScaler


print("=" * 70)
print("FEATURE SCALING USING STANDARD SCALER")
print("=" * 70)


scaler = StandardScaler()


# Fit ONLY on training data
X_train_scaled = scaler.fit_transform(
    X_train
)


# Transform test data
X_test_scaled = scaler.transform(
    X_test
)


print("Scaling completed successfully")


print("\n" + "=" * 70)
print("SHAPE AFTER SCALING")
print("=" * 70)


print(
    f"X_train_scaled shape : {X_train_scaled.shape}"
)

print(
    f"X_test_scaled shape  : {X_test_scaled.shape}"
)

FEATURE SCALING USING STANDARD SCALER
Scaling completed successfully

SHAPE AFTER SCALING
X_train_scaled shape : (242, 13)
X_test_scaled shape  : (61, 13)


In [10]:
# =============================================================================
# STEP 10: CONVERT SCALED DATA BACK TO DATAFRAME
# =============================================================================


X_train_scaled = pd.DataFrame(
    X_train_scaled,
    columns=X_train.columns,
    index=X_train.index
)


X_test_scaled = pd.DataFrame(
    X_test_scaled,
    columns=X_test.columns,
    index=X_test.index
)


print("=" * 70)
print("SCALED TRAIN DATA PREVIEW")
print("=" * 70)


display(
    X_train_scaled.head()
)

SCALED TRAIN DATA PREVIEW


,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal
180,-0.729485,0.68313,0.870169,-0.395692,0.458139,-0.411196,1.022996,0.708371,-0.696177,-0.445445,0.675060,-0.689715,1.179973
208,0.050166,0.68313,-1.184278,-0.054513,0.230598,-0.411196,-0.981579,0.222495,-0.696177,-0.891627,-0.958585,-0.689715,-0.878070
167,-0.061212,-1.46385,-1.184278,0.059213,0.723605,2.431930,1.022996,0.399178,1.436416,-0.891627,-0.958585,0.445734,-0.878070
105,-0.061212,0.68313,-1.184278,-1.305501,1.121803,-0.411196,-0.981579,0.266666,-0.696177,-0.891627,-0.958585,-0.689715,1.179973
297,0.272924,-1.46385,0.870169,0.514117,-0.167601,-0.411196,-0.981579,-1.190962,1.436416,-0.713154,0.675060,-0.689715,1.179973


In [12]:
# =============================================================================
# STEP 11: SCALING VERIFICATION
# =============================================================================

print("=" * 70)
print("MEAN AFTER SCALING")
print("=" * 70)


display(
    X_train_scaled.mean().round(2)
)


print("\n" + "=" * 70)
print("STANDARD DEVIATION AFTER SCALING")
print("=" * 70)


display(
    X_train_scaled.std().round(2)
)

MEAN AFTER SCALING


age        -0.0
sex         0.0
cp          0.0
trestbps   -0.0
chol        0.0
fbs        -0.0
restecg     0.0
thalach    -0.0
exang       0.0
oldpeak     0.0
slope      -0.0
ca          0.0
thal        0.0
dtype: float64


STANDARD DEVIATION AFTER SCALING


age         1.0
sex         1.0
cp          1.0
trestbps    1.0
chol        1.0
fbs         1.0
restecg     1.0
thalach     1.0
exang       1.0
oldpeak     1.0
slope       1.0
ca          1.0
thal        1.0
dtype: float64

# Random Forest

## 1. Baseline

In [14]:
# =============================================================================
# mRMR + RANDOM FOREST (TOP-K = 5,8,10,13)
# =============================================================================

from sklearn.ensemble import RandomForestClassifier


# =============================================================================
# Prepare training data
# =============================================================================

X_train_rf = X_train_scaled.copy()
y_train_rf = pd.Series(
    y_train,
    index=X_train_rf.index,
    name="target"
)


# =============================================================================
# Run Random Forest for maximum K
# =============================================================================

k_values = [5, 8, 10, 13]

max_k = max(k_values)

rf_model = RandomForestClassifier(
    n_estimators=200,
    random_state=42,
    class_weight="balanced"
)

rf_model.fit(
    X_train_rf,
    y_train_rf
)


# =============================================================================
# Create Random Forest ranking
# =============================================================================

rf_scores = pd.DataFrame({
    "Feature": X_train_rf.columns,
    "RF Importance": rf_model.feature_importances_
})

rf_scores = rf_scores.sort_values(
    by="RF Importance",
    ascending=False
).reset_index(drop=True)


# =============================================================================
# Top-K values
# =============================================================================

rf_results = {}


for k in k_values:

    top_features = (
        rf_scores
        .head(k)["Feature"]
        .tolist()
    )


    # Boolean mask for selecting columns
    mask = X_train_scaled.columns.isin(
        top_features
    )


    X_train_selected = X_train_scaled.loc[
        :, mask
    ]

    X_test_selected = X_test_scaled.loc[
        :, mask
    ]


    rf_results[k] = {

        "features": top_features,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "scores": rf_scores.head(k)

    }


    print("=" * 70)
    print(f"TOP {k} RANDOM FOREST FEATURES")
    print("=" * 70)

    print(top_features)

    print("\nImportance:")
    print(rf_scores.head(k))

TOP 5 RANDOM FOREST FEATURES
['thal', 'thalach', 'cp', 'ca', 'age']

Importance:
   Feature  RF Importance
0     thal       0.138340
1  thalach       0.120628
2       cp       0.116574
3       ca       0.105062
4      age       0.096674
TOP 8 RANDOM FOREST FEATURES
['thal', 'thalach', 'cp', 'ca', 'age', 'oldpeak', 'chol', 'trestbps']

Importance:
    Feature  RF Importance
0      thal       0.138340
1   thalach       0.120628
2        cp       0.116574
3        ca       0.105062
4       age       0.096674
5   oldpeak       0.090292
6      chol       0.084277
7  trestbps       0.078291
TOP 10 RANDOM FOREST FEATURES
['thal', 'thalach', 'cp', 'ca', 'age', 'oldpeak', 'chol', 'trestbps', 'exang', 'slope']

Importance:
    Feature  RF Importance
0      thal       0.138340
1   thalach       0.120628
2        cp       0.116574
3        ca       0.105062
4       age       0.096674
5   oldpeak       0.090292
6      chol       0.084277
7  trestbps       0.078291
8     exang       0.051814
9     s

In [15]:
# =============================================================================
# BASELINE MODELS
# =============================================================================

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier


models = {

    "Logistic Regression": LogisticRegression(
        max_iter=1000,
        class_weight="balanced",
        random_state=42
    ),

    "Random Forest": RandomForestClassifier(
        n_estimators=100,
        class_weight="balanced",
        random_state=42
    ),

    "XGBoost": XGBClassifier(
        random_state=42,
        eval_metric="logloss",
        scale_pos_weight=(
            (y_train == 0).sum() /
            (y_train == 1).sum()
        )
    )
}

In [16]:
# =============================================================================
# MODEL EVALUATION — RANDOM FOREST
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)

results_rf = []


for k in k_values:

    Xtr = rf_results[k]["X_train"]

    Xts = rf_results[k]["X_test"]

    features = ", ".join(
        rf_results[k]["features"]
    )


    for model_name, model in models.items():

        model.fit(
            Xtr,
            y_train
        )


        y_pred = model.predict(
            Xts
        )


        y_prob = model.predict_proba(
            Xts
        )[:, 1]


        results_rf.append({

            "Method": "Random Forest",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(
                y_test,
                y_pred
            ),

            "Precision": precision_score(
                y_test,
                y_pred
            ),

            "Recall": recall_score(
                y_test,
                y_pred
            ),

            "F1 Score": f1_score(
                y_test,
                y_pred
            ),

            "ROC-AUC": roc_auc_score(
                y_test,
                y_prob
            )

        })


rf_evaluation = pd.DataFrame(
    results_rf
)

display(rf_evaluation)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,Random Forest,5,Logistic Regression,"thal, thalach, cp, ca, age",0.786885,0.727273,0.857143,0.786885,0.909091
1,Random Forest,5,Random Forest,"thal, thalach, cp, ca, age",0.836066,0.800000,0.857143,0.827586,0.904221
2,Random Forest,5,XGBoost,"thal, thalach, cp, ca, age",0.786885,0.727273,0.857143,0.786885,0.900433
3,Random Forest,8,Logistic Regression,"thal, thalach, cp, ca, age, oldpeak, chol, tre...",0.868852,0.812500,0.928571,0.866667,0.945887
4,Random Forest,8,Random Forest,"thal, thalach, cp, ca, age, oldpeak, chol, tre...",0.885246,0.862069,0.892857,0.877193,0.933983
5,Random Forest,8,XGBoost,"thal, thalach, cp, ca, age, oldpeak, chol, tre...",0.819672,0.793103,0.821429,0.807018,0.896104
6,Random Forest,10,Logistic Regression,"thal, thalach, cp, ca, age, oldpeak, chol, tre...",0.885246,0.818182,0.964286,0.885246,0.949134
7,Random Forest,10,Random Forest,"thal, thalach, cp, ca, age, oldpeak, chol, tre...",0.868852,0.833333,0.892857,0.862069,0.932900
8,Random Forest,10,XGBoost,"thal, thalach, cp, ca, age, oldpeak, chol, tre...",0.852459,0.806452,0.892857,0.847458,0.899351
9,Random Forest,13,Logistic Regression,"thal, thalach, cp, ca, age, oldpeak, chol, tre...",0.868852,0.812500,0.928571,0.866667,0.950216


In [17]:
rf_evaluation.to_csv(
    "../../../results/heart_disease/baseline/rf_baseline_results.csv",
    index=False
)

## 2. DODA

In [18]:
import sys
import doda
from doda import DODASelector
import inspect

print("Python:", sys.executable)
print("DODA:", doda.__file__)
print("Selector:", inspect.getfile(DODASelector))

Python: c:\Users\Anandha Lakshmi\Downloads\Anaconda\python.exe
DODA: c:\Users\Anandha Lakshmi\Downloads\Anaconda\Lib\site-packages\doda\__init__.py
Selector: c:\Users\Anandha Lakshmi\Downloads\Anaconda\Lib\site-packages\doda\selector.py


In [19]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif

from doda import DODASelector

from doda.adapters import SklearnAdapter

from doda.knowledge.providers import JSONProvider

from doda.fusion import RankFusion

In [ ]:
# =============================================================================
# RANDOM FOREST + DODA FEATURE SELECTION
# =============================================================================

import pandas as pd


# -------------------------------------------------------------------------
# Top-K values
# -------------------------------------------------------------------------

k_values = [5, 8, 10, 13]


# -------------------------------------------------------------------------
# Clinical knowledge provider
# -------------------------------------------------------------------------

provider = JSONProvider(
    "heartdisease_weights.json"
)


# -------------------------------------------------------------------------
# Fusion method
# -------------------------------------------------------------------------

fusion = RankFusion()


# -------------------------------------------------------------------------
# Store DODA results
# -------------------------------------------------------------------------

rf_doda_results = {}


# =============================================================================
# RUN RANDOM FOREST + DODA FOR EACH TOP-K
# =============================================================================

for k in k_values:

    print("\n" + "=" * 70)
    print(f"RANDOM FOREST + DODA — TOP {k} FEATURES")
    print("=" * 70)


    # ---------------------------------------------------------------------
    # Random Forest operator
    # ---------------------------------------------------------------------

    rf_operator = SklearnAdapter(
        RandomForestClassifier(
            n_estimators=200,
            random_state=42,
            class_weight="balanced"
        )
    )


    # ---------------------------------------------------------------------
    # DODA selector
    # ---------------------------------------------------------------------

    selector = DODASelector(

        operators=[rf_operator],

        provider=provider,

        fusion=fusion,

        top_k=k
    )


    # ---------------------------------------------------------------------
    # Fit DODA on TRAINING data
    # ---------------------------------------------------------------------

    X_train_selected = selector.fit_transform(

        X_train_scaled,

        y_train

    )


    # ---------------------------------------------------------------------
    # Transform TEST data
    # ---------------------------------------------------------------------

    X_test_selected = selector.transform(

        X_test_scaled

    )


    # ---------------------------------------------------------------------
    # Selected features
    # ---------------------------------------------------------------------

    selected_features = (
        selector.get_selected_features()
    )


    print("\nSelected Features:")
    print(selected_features)


    # ---------------------------------------------------------------------
    # Store everything needed for evaluation
    # ---------------------------------------------------------------------

    rf_doda_results[k] = {

        "selector": selector,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "features": selected_features,

        "raw_math_scores":
            selector.raw_math_scores_,

        "math_scores":
            selector.math_scores_,

        "clinical_weights":
            selector.clinical_weights_,

        "final_scores":
            selector.final_scores_
    }


print("\n" + "=" * 70)
print("RANDOM FOREST + DODA FEATURE SELECTION COMPLETED")
print("=" * 70)


RANDOM FOREST + DODA — TOP 5 FEATURES
Provider inside knowledge engine: <doda.knowledge.providers.json_provider.JSONProvider object at 0x000001777533CAD0>
DODA Pipeline Started
Running: RandomForestClassifier
DODA feature order:
['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak']
Number of features: 13
Number of feature importances: 13
Operator Scores:
{'RandomForestClassifier_1': {'age': np.float64(0.0966735741067629), 'sex': np.float64(0.04097798577342465), 'cp': np.float64(0.1165744623287366), 'trestbps': np.float64(0.07829131674305914), 'chol': np.float64(0.08427654782893294), 'fbs': np.float64(0.01137173061045471), 'restecg': np.float64(0.022061148446768607), 'thalach': np.float64(0.1206280324818276), 'exang': np.float64(0.051814045539448746), 'oldpeak': np.float64(0.09029189464775904), 'slope': np.float64(0.04363740284934486), 'ca': np.float64(0.10506181498012705), 'thal': np.float64(0.1383400436633532)}}
Resolving scores from: 1 operators



In [21]:
# =============================================================================
# INSPECT MATHEMATICAL SCORES
# =============================================================================

k = 10

selector = rf_doda_results[k]["selector"]


# -------------------------------------------------------------------------
# Raw mathematical scores
# -------------------------------------------------------------------------

raw_math_scores = pd.DataFrame(
    selector.raw_math_scores_.items(),
    columns=[
        "Feature",
        "Raw Math Score"
    ]
)


# -------------------------------------------------------------------------
# Normalized mathematical scores
# -------------------------------------------------------------------------

normalized_math_scores = pd.DataFrame(
    selector.math_scores_.items(),
    columns=[
        "Feature",
        "Normalized Math Score"
    ]
)


# -------------------------------------------------------------------------
# Combine both
# -------------------------------------------------------------------------

math_scores = raw_math_scores.merge(
    normalized_math_scores,
    on="Feature"
)


math_scores = math_scores.sort_values(
    by="Normalized Math Score",
    ascending=False
)


print("=" * 70)
print("MATHEMATICAL SCORES")
print("=" * 70)

display(math_scores)

MATHEMATICAL SCORES


,Feature,Raw Math Score,Normalized Math Score
12,thal,0.138340,1.000000
7,thalach,0.120628,0.871968
2,cp,0.116574,0.842666
11,ca,0.105062,0.759446
0,age,0.096674,0.698811
9,oldpeak,0.090292,0.652681
4,chol,0.084277,0.609199
3,trestbps,0.078291,0.565934
8,exang,0.051814,0.374541
10,slope,0.043637,0.315436


In [22]:
# =============================================================================
# INSPECT CLINICAL WEIGHTS
# =============================================================================

clinical_weights = pd.DataFrame(
    selector.clinical_weights_.items(),
    columns=[
        "Feature",
        "Clinical Weight"
    ]
)


clinical_weights = clinical_weights.sort_values(
    by="Clinical Weight",
    ascending=False
)


print("=" * 70)
print("CLINICAL WEIGHTS")
print("=" * 70)

display(clinical_weights)

CLINICAL WEIGHTS


,Feature,Clinical Weight
2,cp,1.0
9,oldpeak,1.0
11,ca,1.0
12,thal,1.0
8,exang,1.0
10,slope,0.9
0,age,0.9
3,trestbps,0.9
4,chol,0.8
7,thalach,0.8


In [23]:
# =============================================================================
# INSPECT FINAL RANK FUSION SCORES
# =============================================================================

final_scores = pd.DataFrame(
    selector.final_scores_.items(),
    columns=[
        "Feature",
        "Final Rank Fusion Score"
    ]
)


final_scores = final_scores.sort_values(
    by="Final Rank Fusion Score",
    ascending=False
)


print("=" * 70)
print("FINAL RANK FUSION SCORES")
print("=" * 70)

display(final_scores)

FINAL RANK FUSION SCORES


,Feature,Final Rank Fusion Score
12,thal,0.032787
2,cp,0.032266
11,ca,0.032018
7,thalach,0.032002
9,oldpeak,0.031545
0,age,0.031514
8,exang,0.030886
3,trestbps,0.030835
4,chol,0.030798
10,slope,0.030415


In [24]:
# =============================================================================
# SELECTED FEATURES
# =============================================================================

selector.get_selected_features()

['thal',
 'cp',
 'ca',
 'thalach',
 'oldpeak',
 'age',
 'exang',
 'trestbps',
 'chol',
 'slope']

In [25]:
# =============================================================================
# COMPARE RANKINGS
# =============================================================================

comparison = selector.compare_scores()

display(comparison)


MATHEMATICAL vs CLINICAL vs RANKFUSION

Fusion Method: RankFusion

Top Features After Fusion:
    Feature  Math Rank  Clinical Rank  Final Rank  Rank Change
0      thal          1              5           1            0
1        cp          3              1           2            1
2        ca          4              4           3            1
3   thalach          2             11           4           -2
4   oldpeak          6              3           5            1
5       age          5              6           6           -1
6     exang          9              2           7            2
7  trestbps          8              7           8            0
8      chol          7              9           9           -2
9     slope         10              8          10            0


,Feature,Math Rank,Normalized Math Score,Clinical Rank,Clinical Weight,Final Rank,Final Score,Rank Change
0,thal,1,1.0000,5,1.0,1,0.0328,0
1,cp,3,0.8427,1,1.0,2,0.0323,1
2,ca,4,0.7594,4,1.0,3,0.0320,1
3,thalach,2,0.8720,11,0.8,4,0.0320,-2
4,oldpeak,6,0.6527,3,1.0,5,0.0315,1
5,age,5,0.6988,6,0.9,6,0.0315,-1
6,exang,9,0.3745,2,1.0,7,0.0309,2
7,trestbps,8,0.5659,7,0.9,8,0.0308,0
8,chol,7,0.6092,9,0.8,9,0.0308,-2
9,slope,10,0.3154,8,0.9,10,0.0304,0


In [26]:
# =============================================================================
#  MODELS
# =============================================================================

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier


models = {

    "Logistic Regression":
        LogisticRegression(
            max_iter=1000,
            random_state=42
        ),


    "Random Forest":
        RandomForestClassifier(
            n_estimators=200,
            random_state=42
        ),


    "XGBoost":
        XGBClassifier(
            n_estimators=200,
            random_state=42,
            eval_metric="logloss"
        )
}

In [30]:
# =============================================================================
# RANK FUSION DODA MODEL EVALUATION — RF
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)


rf_doda_results_eval = []


for k in k_values:

    Xtr = rf_doda_results[k]["X_train"]

    Xts = rf_doda_results[k]["X_test"]

    features = ", ".join(
        rf_doda_results[k]["features"]
    )


    for model_name, model in models.items():

        model.fit(
            Xtr,
            y_train
        )


        y_pred = model.predict(
            Xts
        )


        y_prob = model.predict_proba(
            Xts
        )[:, 1]


        rf_doda_results_eval.append({

            "Method": "Random Forest + Rank Fusion DODA",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(
                y_test,
                y_pred
            ),

            "Precision": precision_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "Recall": recall_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "F1 Score": f1_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "ROC-AUC": roc_auc_score(
                y_test,
                y_prob
            )

        })


rf_doda_evaluation = pd.DataFrame(
    rf_doda_results_eval
)


display(rf_doda_evaluation)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,Random Forest + Rank Fusion DODA,5,Logistic Regression,"thal, cp, ca, thalach, oldpeak",0.868852,0.812500,0.928571,0.866667,0.933983
1,Random Forest + Rank Fusion DODA,5,Random Forest,"thal, cp, ca, thalach, oldpeak",0.819672,0.793103,0.821429,0.807018,0.917208
2,Random Forest + Rank Fusion DODA,5,XGBoost,"thal, cp, ca, thalach, oldpeak",0.786885,0.741935,0.821429,0.779661,0.889610
3,Random Forest + Rank Fusion DODA,8,Logistic Regression,"thal, cp, ca, thalach, oldpeak, age, exang, tr...",0.852459,0.806452,0.892857,0.847458,0.953463
4,Random Forest + Rank Fusion DODA,8,Random Forest,"thal, cp, ca, thalach, oldpeak, age, exang, tr...",0.852459,0.806452,0.892857,0.847458,0.952381
5,Random Forest + Rank Fusion DODA,8,XGBoost,"thal, cp, ca, thalach, oldpeak, age, exang, tr...",0.754098,0.685714,0.857143,0.761905,0.916667
6,Random Forest + Rank Fusion DODA,10,Logistic Regression,"thal, cp, ca, thalach, oldpeak, age, exang, tr...",0.868852,0.812500,0.928571,0.866667,0.948052
7,Random Forest + Rank Fusion DODA,10,Random Forest,"thal, cp, ca, thalach, oldpeak, age, exang, tr...",0.885246,0.838710,0.928571,0.881356,0.943182
8,Random Forest + Rank Fusion DODA,10,XGBoost,"thal, cp, ca, thalach, oldpeak, age, exang, tr...",0.836066,0.781250,0.892857,0.833333,0.916667
9,Random Forest + Rank Fusion DODA,13,Logistic Regression,"thal, cp, ca, thalach, oldpeak, age, exang, tr...",0.868852,0.812500,0.928571,0.866667,0.951299


In [31]:
# =============================================================================
# SAVE DODA RESULTS
# =============================================================================


rf_doda_evaluation.to_csv(
    "../../../results/heart_disease/doda/rf_doda_results.csv",
    index=False
)

In [32]:
# =============================================================================
# COMBINED COMPARISON
# =============================================================================


comparison_df = pd.concat(
    [
        rf_evaluation,
        rf_doda_evaluation
    ],
    ignore_index=True
)


display(
    comparison_df
)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,Random Forest,5,Logistic Regression,"thal, thalach, cp, ca, age",0.786885,0.727273,0.857143,0.786885,0.909091
1,Random Forest,5,Random Forest,"thal, thalach, cp, ca, age",0.836066,0.800000,0.857143,0.827586,0.904221
2,Random Forest,5,XGBoost,"thal, thalach, cp, ca, age",0.786885,0.727273,0.857143,0.786885,0.900433
3,Random Forest,8,Logistic Regression,"thal, thalach, cp, ca, age, oldpeak, chol, tre...",0.868852,0.812500,0.928571,0.866667,0.945887
4,Random Forest,8,Random Forest,"thal, thalach, cp, ca, age, oldpeak, chol, tre...",0.885246,0.862069,0.892857,0.877193,0.933983
5,Random Forest,8,XGBoost,"thal, thalach, cp, ca, age, oldpeak, chol, tre...",0.819672,0.793103,0.821429,0.807018,0.896104
6,Random Forest,10,Logistic Regression,"thal, thalach, cp, ca, age, oldpeak, chol, tre...",0.885246,0.818182,0.964286,0.885246,0.949134
7,Random Forest,10,Random Forest,"thal, thalach, cp, ca, age, oldpeak, chol, tre...",0.868852,0.833333,0.892857,0.862069,0.932900
8,Random Forest,10,XGBoost,"thal, thalach, cp, ca, age, oldpeak, chol, tre...",0.852459,0.806452,0.892857,0.847458,0.899351
9,Random Forest,13,Logistic Regression,"thal, thalach, cp, ca, age, oldpeak, chol, tre...",0.868852,0.812500,0.928571,0.866667,0.950216


In [33]:
# ============================================================
# RANDOM FOREST + DODA STABILITY ANALYSIS — ALL TOP-K VALUES
# Heart Disease Dataset
# ============================================================


import pandas as pd
import numpy as np

from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.impute import SimpleImputer
from sklearn.ensemble import RandomForestClassifier

from doda import DODASelector
from doda.knowledge import JSONProvider
from doda.fusion import RankFusion
from doda.adapters.sklearn_adapter import SklearnAdapter


# ============================================================
# CONFIGURATION
# ============================================================

TOP_K_VALUES = [5, 8, 10, 13]

N_SPLITS = 5
N_REPEATS = 5
RANDOM_STATE = 42

WEIGHTS_FILE = "heartdisease_weights.json"


# ============================================================
# DATA
# ============================================================

# X and y should contain the ORIGINAL Heart Disease data
# before imputation/scaling.
#
# Example:
#
# X = df.drop(columns=["target"])
# y = df["target"]


# ============================================================
# REPEATED STRATIFIED CV
# ============================================================

cv = RepeatedStratifiedKFold(
    n_splits=N_SPLITS,
    n_repeats=N_REPEATS,
    random_state=RANDOM_STATE
)


# ============================================================
# STORAGE
# ============================================================

all_runs = []
all_jaccard = []
all_frequency = []


# ============================================================
# LOOP THROUGH ALL TOP-K VALUES
# ============================================================

for TOP_K in TOP_K_VALUES:

    print("\n" + "=" * 70)
    print(f"RANDOM FOREST + DODA STABILITY ANALYSIS — TOP-{TOP_K}")
    print("=" * 70)

    selected_feature_sets = []

    run_number = 0


    # ========================================================
    # 25 REPEATED CV RUNS
    # ========================================================

    for train_idx, val_idx in cv.split(X, y):

        run_number += 1

        X_train = X.iloc[train_idx].copy()
        y_train = y.iloc[train_idx].copy()

        X_val = X.iloc[val_idx].copy()
        y_val = y.iloc[val_idx].copy()


        # ----------------------------------------------------
        # IMPUTATION
        # ----------------------------------------------------

        imputer = SimpleImputer(
            strategy="median"
        )

        X_train_imp = pd.DataFrame(
            imputer.fit_transform(X_train),
            columns=X_train.columns,
            index=X_train.index
        )

        X_val_imp = pd.DataFrame(
            imputer.transform(X_val),
            columns=X_val.columns,
            index=X_val.index
        )


        # ----------------------------------------------------
        # RANDOM FOREST OPERATOR
        # ----------------------------------------------------

        rf_operator = SklearnAdapter(
            RandomForestClassifier(
                n_estimators=200,
                random_state=RANDOM_STATE,
                class_weight="balanced"
            )
        )


        # ----------------------------------------------------
        # CLINICAL KNOWLEDGE PROVIDER
        # ----------------------------------------------------

        provider = JSONProvider(
            WEIGHTS_FILE
        )


        # ----------------------------------------------------
        # DODA SELECTOR
        # ----------------------------------------------------

        doda_selector = DODASelector(

            operators=[rf_operator],

            provider=provider,

            fusion=RankFusion(),

            top_k=TOP_K

        )


        # ----------------------------------------------------
        # FIT ONLY ON TRAINING FOLD
        # ----------------------------------------------------

        doda_selector.fit(
            X_train_imp,
            y_train
        )


        # ----------------------------------------------------
        # SELECTED FEATURES
        # ----------------------------------------------------

        selected_features = list(
            doda_selector.get_selected_features()
        )


        selected_feature_sets.append(
            set(selected_features)
        )


        all_runs.append({

            "Top_K": TOP_K,

            "Run": run_number,

            "Selected_Features": selected_features

        })


        print(
            f"Run {run_number:02d}: "
            f"{selected_features}"
        )


    # ========================================================
    # JACCARD SIMILARITY
    # ========================================================

    jaccard_scores = []


    for i in range(
        len(selected_feature_sets)
    ):

        for j in range(
            i + 1,
            len(selected_feature_sets)
        ):

            set_a = selected_feature_sets[i]
            set_b = selected_feature_sets[j]


            intersection = len(
                set_a.intersection(set_b)
            )

            union = len(
                set_a.union(set_b)
            )


            jaccard = (
                intersection / union
            )


            jaccard_scores.append(
                jaccard
            )


            all_jaccard.append({

                "Top_K": TOP_K,

                "Run_A": i + 1,

                "Run_B": j + 1,

                "Jaccard_Similarity": jaccard

            })


    # ========================================================
    # STABILITY SUMMARY
    # ========================================================

    mean_jaccard = np.mean(
        jaccard_scores
    )

    std_jaccard = np.std(
        jaccard_scores
    )

    min_jaccard = np.min(
        jaccard_scores
    )

    max_jaccard = np.max(
        jaccard_scores
    )


    print("\n" + "-" * 70)
    print(f"TOP-{TOP_K} STABILITY SUMMARY")
    print("-" * 70)

    print(
        f"Number of runs:          "
        f"{len(selected_feature_sets)}"
    )

    print(
        f"Number of comparisons:   "
        f"{len(jaccard_scores)}"
    )

    print(
        f"Mean Jaccard Similarity: "
        f"{mean_jaccard:.4f}"
    )

    print(
        f"Standard Deviation:      "
        f"{std_jaccard:.4f}"
    )

    print(
        f"Minimum Jaccard:         "
        f"{min_jaccard:.4f}"
    )

    print(
        f"Maximum Jaccard:         "
        f"{max_jaccard:.4f}"
    )


    # ========================================================
    # FEATURE SELECTION FREQUENCY
    # ========================================================

    feature_counts = {}


    for feature_set in selected_feature_sets:

        for feature in feature_set:

            feature_counts[feature] = (
                feature_counts.get(feature, 0) + 1
            )


    for feature, count in feature_counts.items():

        all_frequency.append({

            "Top_K": TOP_K,

            "Feature": feature,

            "Selection_Count": count,

            "Selection_Frequency":
                count / len(selected_feature_sets)

        })


    frequency_df_topk = pd.DataFrame(

        [

            {

                "Feature": feature,

                "Selection_Count": count,

                "Selection_Frequency":
                    count / len(selected_feature_sets)

            }

            for feature, count
            in feature_counts.items()

        ]

    ).sort_values(

        "Selection_Frequency",

        ascending=False

    )


    print(
        "\nFEATURE SELECTION FREQUENCY"
    )

    print(
        frequency_df_topk.to_string(
            index=False
        )
    )


# ============================================================
# CREATE DATAFRAMES
# ============================================================

runs_df = pd.DataFrame(
    all_runs
)

jaccard_df = pd.DataFrame(
    all_jaccard
)

frequency_df = pd.DataFrame(
    all_frequency
)


# ============================================================
# SUMMARY TABLE
# ============================================================

summary_rows = []


for TOP_K in TOP_K_VALUES:

    subset = jaccard_df[
        jaccard_df["Top_K"] == TOP_K
    ]["Jaccard_Similarity"]


    summary_rows.append({

        "Top_K": TOP_K,

        "Mean_Jaccard":
            subset.mean(),

        "Std_Jaccard":
            subset.std(),

        "Min_Jaccard":
            subset.min(),

        "Max_Jaccard":
            subset.max()

    })


summary_df = pd.DataFrame(
    summary_rows
)


# ============================================================
# DISPLAY FINAL SUMMARY
# ============================================================

print("\n" + "=" * 70)
print("RANDOM FOREST + DODA STABILITY — ALL TOP-K VALUES")
print("=" * 70)

print(

    summary_df.to_string(

        index=False,

        float_format=lambda x:
            f"{x:.4f}"

    )

)


# ============================================================
# SAVE RESULTS
# ============================================================

runs_df.to_csv(
    "../../../results/heart_disease/"
    "rf_doda_stability_all_topk_runs.csv",
    index=False
)

jaccard_df.to_csv(
    "../../../results/heart_disease/"
    "rf_doda_stability_all_topk_jaccard.csv",
    index=False
)

frequency_df.to_csv(
    "../../../results/heart_disease/"
    "rf_doda_stability_all_topk_frequency.csv",
    index=False
)

summary_df.to_csv(
    "../../../results/heart_disease/"
    "rf_doda_stability_all_topk_summary.csv",
    index=False
)


print("\nFiles saved:")

print(
    "1. rf_doda_stability_all_topk_runs.csv"
)

print(
    "2. rf_doda_stability_all_topk_jaccard.csv"
)

print(
    "3. rf_doda_stability_all_topk_frequency.csv"
)

print(
    "4. rf_doda_stability_all_topk_summary.csv"
)


RANDOM FOREST + DODA STABILITY ANALYSIS — TOP-5
Provider inside knowledge engine: <doda.knowledge.providers.json_provider.JSONProvider object at 0x00000177760FB250>
DODA Pipeline Started
Running: RandomForestClassifier
DODA feature order:
['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak']
Number of features: 13
Number of feature importances: 13
Operator Scores:
{'RandomForestClassifier_1': {'age': np.float64(0.09662229851782783), 'sex': np.float64(0.0379793873855365), 'cp': np.float64(0.11799954200094252), 'trestbps': np.float64(0.07674413347867001), 'chol': np.float64(0.08691380790290808), 'fbs': np.float64(0.01118902301109723), 'restecg': np.float64(0.022949682074261193), 'thalach': np.float64(0.12152056409551298), 'exang': np.float64(0.0609066282299226), 'oldpeak': np.float64(0.1056730494553615), 'slope': np.float64(0.039139363488592045), 'ca': np.float64(0.10832103250099333), 'thal': np.float64(0.11404148785837412)}}
Resolving scores from: 1 

In [34]:
# =============================================================================
# REPEATED STRATIFIED CROSS-VALIDATION
# RANDOM FOREST VS DODA-RANDOM FOREST WITH RANK FUSION
# =============================================================================

import numpy as np
import pandas as pd

from sklearn.base import clone
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)

from doda import DODASelector
from doda.adapters.sklearn_adapter import SklearnAdapter
from doda.knowledge import JSONProvider
from doda.fusion import RankFusion


# =============================================================================
# CONFIGURATION
# =============================================================================

TOP_K_VALUES = [5, 8, 10, 13]

N_SPLITS = 5
N_REPEATS = 5
RANDOM_STATE = 42

target_column = "target"

X = df.drop(columns=[target_column])
y = df[target_column]

feature_names = X.columns.tolist()


# =============================================================================
# MODELS
# =============================================================================

models = {

    "Logistic Regression": LogisticRegression(
        max_iter=2000,
        random_state=RANDOM_STATE
    ),

    "Random Forest": RandomForestClassifier(
        n_estimators=300,
        random_state=RANDOM_STATE,
        n_jobs=-1
    ),

    "XGBoost": XGBClassifier(
        n_estimators=300,
        max_depth=3,
        learning_rate=0.05,
        subsample=0.8,
        colsample_bytree=0.8,
        eval_metric="logloss",
        random_state=RANDOM_STATE,
        n_jobs=-1
    )
}


# =============================================================================
# CLINICAL KNOWLEDGE PROVIDER
# =============================================================================

provider = JSONProvider(
    "heartdisease_weights.json"
)


# =============================================================================
# REPEATED STRATIFIED CROSS-VALIDATION
# =============================================================================

cv = RepeatedStratifiedKFold(
    n_splits=N_SPLITS,
    n_repeats=N_REPEATS,
    random_state=RANDOM_STATE
)


# =============================================================================
# RESULTS STORAGE
# =============================================================================

results = []


# =============================================================================
# CROSS-VALIDATION LOOP
# =============================================================================

for run_number, (train_idx, test_idx) in enumerate(
    cv.split(X, y),
    start=1
):

    print("\n" + "=" * 80)
    print(f"REPEATED STRATIFIED CV RUN {run_number}")
    print("=" * 80)

    X_train = X.iloc[train_idx].copy()
    X_test = X.iloc[test_idx].copy()

    y_train = y.iloc[train_idx].copy()
    y_test = y.iloc[test_idx].copy()


    # =========================================================================
    # IMPUTATION
    # =========================================================================

    imputer = SimpleImputer(
        strategy="median"
    )

    X_train_imp_array = imputer.fit_transform(
        X_train
    )

    X_test_imp_array = imputer.transform(
        X_test
    )

    X_train_imp = pd.DataFrame(
        X_train_imp_array,
        columns=feature_names,
        index=X_train.index
    )

    X_test_imp = pd.DataFrame(
        X_test_imp_array,
        columns=feature_names,
        index=X_test.index
    )


    # =========================================================================
    # FEATURE SELECTION FOR EACH TOP-K VALUE
    # =========================================================================

    for top_k in TOP_K_VALUES:

        print("\n" + "-" * 80)
        print(f"TOP-K = {top_k}")
        print("-" * 80)


        # =====================================================================
        # BASELINE RANDOM FOREST FEATURE SELECTION
        # =====================================================================

        rf_operator_baseline = RandomForestClassifier(
            n_estimators=300,
            random_state=RANDOM_STATE,
            n_jobs=-1
        )

        rf_operator_baseline.fit(
            X_train_imp,
            y_train
        )

        rf_importance = pd.Series(
            rf_operator_baseline.feature_importances_,
            index=X_train_imp.columns
        )

        rf_features = (
            rf_importance
            .sort_values(
                ascending=False
            )
            .head(top_k)
            .index
            .tolist()
        )

        print("Random Forest-selected features:")
        print(rf_features)


        # =====================================================================
        # DODA-RANDOM FOREST FEATURE SELECTION
        # =====================================================================

        rf_operator = SklearnAdapter(
            RandomForestClassifier(
                n_estimators=300,
                random_state=RANDOM_STATE,
                n_jobs=-1
            )
        )

        doda_selector = DODASelector(
            operators=[rf_operator],
            provider=provider,
            fusion=RankFusion(),
            top_k=top_k
        )

        doda_selector.fit(
            X_train_imp,
            y_train
        )

        doda_features = list(
            doda_selector.get_selected_features()
        )

        print("DODA-selected features:")
        print(doda_features)


        # =====================================================================
        # PREPARE SELECTED FEATURE MATRICES
        # =====================================================================

        X_train_rf_selected = (
            X_train_imp[rf_features]
        )

        X_test_rf_selected = (
            X_test_imp[rf_features]
        )

        X_train_doda_selected = (
            X_train_imp[doda_features]
        )

        X_test_doda_selected = (
            X_test_imp[doda_features]
        )


        # =====================================================================
        # MODEL EVALUATION
        # =====================================================================

        selected_feature_sets = {

            "Random Forest": (
                X_train_rf_selected,
                X_test_rf_selected,
                rf_features
            ),

            "DODA": (
                X_train_doda_selected,
                X_test_doda_selected,
                doda_features
            )

        }


        for method, (
            X_train_selected,
            X_test_selected,
            selected_features
        ) in selected_feature_sets.items():

            for model_name, model_template in models.items():

                model = clone(
                    model_template
                )

                X_train_model = (
                    X_train_selected.copy()
                )

                X_test_model = (
                    X_test_selected.copy()
                )


                # -------------------------------------------------------------
                # Scale only Logistic Regression
                # -------------------------------------------------------------

                if model_name == "Logistic Regression":

                    model_scaler = StandardScaler()

                    X_train_model = (
                        model_scaler.fit_transform(
                            X_train_model
                        )
                    )

                    X_test_model = (
                        model_scaler.transform(
                            X_test_model
                        )
                    )


                # -------------------------------------------------------------
                # Train model
                # -------------------------------------------------------------

                model.fit(
                    X_train_model,
                    y_train
                )

                y_pred = model.predict(
                    X_test_model
                )

                y_prob = model.predict_proba(
                    X_test_model
                )[:, 1]


                # -------------------------------------------------------------
                # Calculate metrics
                # -------------------------------------------------------------

                accuracy = accuracy_score(
                    y_test,
                    y_pred
                )

                precision = precision_score(
                    y_test,
                    y_pred,
                    zero_division=0
                )

                recall = recall_score(
                    y_test,
                    y_pred,
                    zero_division=0
                )

                f1 = f1_score(
                    y_test,
                    y_pred,
                    zero_division=0
                )

                roc_auc = roc_auc_score(
                    y_test,
                    y_prob
                )


                # -------------------------------------------------------------
                # Store results
                # -------------------------------------------------------------

                results.append({

                    "Run": run_number,

                    "Method": method,

                    "Top_K": top_k,

                    "Model": model_name,

                    "Accuracy": accuracy,

                    "Precision": precision,

                    "Recall": recall,

                    "F1": f1,

                    "ROC_AUC": roc_auc,

                    "Selected_Features": selected_features

                })


# =============================================================================
# RESULTS DATAFRAME
# =============================================================================

results_df = pd.DataFrame(
    results
)


# =============================================================================
# DISPLAY RESULTS
# =============================================================================

print("\n" + "=" * 80)
print("REPEATED STRATIFIED CV RESULTS — RANDOM FOREST VS DODA-RANDOM FOREST")
print("=" * 80)

display(
    results_df.head()
)

print(
    "Total evaluation rows:",
    len(results_df)
)


REPEATED STRATIFIED CV RUN 1

--------------------------------------------------------------------------------
TOP-K = 5
--------------------------------------------------------------------------------
Random Forest-selected features:
['cp', 'ca', 'thalach', 'oldpeak', 'thal']
Provider inside knowledge engine: <doda.knowledge.providers.json_provider.JSONProvider object at 0x00000177762BC650>
DODA Pipeline Started
Running: RandomForestClassifier
DODA feature order:
['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak']
Number of features: 13
Number of feature importances: 13
Operator Scores:
{'RandomForestClassifier_1': {'age': np.float64(0.09917931305836519), 'sex': np.float64(0.03664066137597381), 'cp': np.float64(0.1184469104724246), 'trestbps': np.float64(0.07833053868689366), 'chol': np.float64(0.08250349006191013), 'fbs': np.float64(0.011054763406595183), 'restecg': np.float64(0.02387904296865469), 'thalach': np.float64(0.11307957339465466), 'ex

,Run,Method,Top_K,Model,Accuracy,Precision,Recall,F1,ROC_AUC,Selected_Features
0,1,Random Forest,5,Logistic Regression,0.934426,0.928571,0.928571,0.928571,0.952381,"[cp, ca, thalach, oldpeak, thal]"
1,1,Random Forest,5,Random Forest,0.852459,0.880000,0.785714,0.830189,0.934524,"[cp, ca, thalach, oldpeak, thal]"
2,1,Random Forest,5,XGBoost,0.852459,0.880000,0.785714,0.830189,0.903680,"[cp, ca, thalach, oldpeak, thal]"
3,1,DODA,5,Logistic Regression,0.934426,0.928571,0.928571,0.928571,0.952381,"[cp, ca, oldpeak, thal, thalach]"
4,1,DODA,5,Random Forest,0.852459,0.880000,0.785714,0.830189,0.927489,"[cp, ca, oldpeak, thal, thalach]"


Total evaluation rows: 600


In [35]:
# =============================================================================
# PREPARE FEATURE SETS FOR STABILITY ANALYSIS
# mRMR VS DODA-mRMR
# =============================================================================

import ast
import numpy as np
import pandas as pd


# =============================================================================
# CONVERT SELECTED FEATURES TO SET
# =============================================================================

def convert_to_set(features):
    """
    Convert Selected_Features into a Python set.
    Handles lists, tuples, NumPy arrays, and strings.
    """

    if isinstance(features, str):

        # Handle string representation of a list
        if features.startswith("[") and features.endswith("]"):

            try:
                features = ast.literal_eval(features)

            except Exception:
                features = features.split(",")

        else:
            features = features.split(",")


    return {
        str(feature).strip().strip("'").strip('"')
        for feature in features
        if str(feature).strip()
    }


# =============================================================================
# USE ONLY ONE MODEL
# =============================================================================

# Feature selection is independent of the predictive model.
# Therefore, Logistic Regression rows are sufficient.

stability_df = results_df[
    results_df["Model"] == "Logistic Regression"
].copy()


# =============================================================================
# CONVERT FEATURE LISTS TO SETS
# =============================================================================

stability_df["Feature_Set"] = (
    stability_df["Selected_Features"]
    .apply(convert_to_set)
)


# =============================================================================
# CHECK RESULTS
# =============================================================================

print("=" * 80)
print("FEATURE SETS PREPARED FOR STABILITY ANALYSIS")
print("=" * 80)

print(
    "Rows used for stability analysis:",
    len(stability_df)
)


# =============================================================================
# DISPLAY SAMPLE
# =============================================================================

display(
    stability_df[
        [
            "Run",
            "Method",
            "Top_K",
            "Feature_Set"
        ]
    ].head(10)
)


# =============================================================================
# CHECK FEATURE COUNT
# =============================================================================

stability_df["Feature_Count"] = (
    stability_df["Feature_Set"]
    .apply(len)
)


print("\nFeature count check:")

display(
    stability_df[
        [
            "Method",
            "Top_K",
            "Feature_Count"
        ]
    ]
    .drop_duplicates()
    .sort_values(
        ["Method", "Top_K"]
    )
)


# =============================================================================
# VERIFY EXPECTED TOP-K
# =============================================================================

invalid_rows = stability_df[
    stability_df["Feature_Count"] != stability_df["Top_K"]
]


if len(invalid_rows) == 0:

    print(
        "\n✓ All feature sets contain the expected number of features."
    )

else:

    print(
        "\n⚠ WARNING: Some feature sets do not match their Top-K value."
    )

    display(
        invalid_rows[
            [
                "Run",
                "Method",
                "Top_K",
                "Feature_Count",
                "Feature_Set"
            ]
        ]
    )

FEATURE SETS PREPARED FOR STABILITY ANALYSIS
Rows used for stability analysis: 200


,Run,Method,Top_K,Feature_Set
0,1,Random Forest,5,"{thal, cp, oldpeak, ca, thalach}"
3,1,DODA,5,"{thal, cp, oldpeak, ca, thalach}"
6,1,Random Forest,8,"{thal, age, trestbps, cp, chol, oldpeak, ca, t..."
9,1,DODA,8,"{thal, age, trestbps, cp, oldpeak, exang, ca, ..."
12,1,Random Forest,10,"{thal, age, trestbps, cp, chol, slope, oldpeak..."
15,1,DODA,10,"{thal, age, trestbps, cp, chol, slope, oldpeak..."
18,1,Random Forest,13,"{thal, age, trestbps, fbs, cp, chol, slope, ol..."
21,1,DODA,13,"{thal, age, trestbps, fbs, cp, chol, slope, ol..."
24,2,Random Forest,5,"{thal, cp, oldpeak, ca, thalach}"
27,2,DODA,5,"{thal, cp, oldpeak, ca, thalach}"



Feature count check:


,Method,Top_K,Feature_Count
3,DODA,5,5
9,DODA,8,8
15,DODA,10,10
21,DODA,13,13
0,Random Forest,5,5
6,Random Forest,8,8
12,Random Forest,10,10
18,Random Forest,13,13



✓ All feature sets contain the expected number of features.


In [36]:
# =============================================================================
# TEST 1: FEATURE SELECTION FREQUENCY
# RF VS DODA-RF
# =============================================================================

selection_frequency = []


# =============================================================================
# LOOP THROUGH METHODS AND TOP-K VALUES
# =============================================================================

for method in stability_df["Method"].unique():

    for top_k in sorted(
        stability_df["Top_K"].unique()
    ):

        subset = stability_df[
            (stability_df["Method"] == method) &
            (stability_df["Top_K"] == top_k)
        ]


        # -------------------------------------------------------------
        # Number of CV runs
        # -------------------------------------------------------------

        total_runs = len(subset)


        # -------------------------------------------------------------
        # All features appearing across the 25 runs
        # -------------------------------------------------------------

        all_features = sorted(
            set().union(
                *subset["Feature_Set"]
            )
        )


        # -------------------------------------------------------------
        # Calculate selection frequency
        # -------------------------------------------------------------

        for feature in all_features:

            selected_count = sum(
                feature in feature_set
                for feature_set
                in subset["Feature_Set"]
            )


            selection_frequency.append({

                "Method": method,

                "Top_K": top_k,

                "Feature": feature,

                "Selected_Count": selected_count,

                "Total_Runs": total_runs,

                "Selection_Frequency_Percentage":
                    (
                        selected_count /
                        total_runs
                    ) * 100

            })


# =============================================================================
# CREATE DATAFRAME
# =============================================================================

selection_frequency_df = pd.DataFrame(
    selection_frequency
)


# =============================================================================
# SORT RESULTS
# =============================================================================

selection_frequency_df = (
    selection_frequency_df
    .sort_values(
        by=[
            "Method",
            "Top_K",
            "Selection_Frequency_Percentage"
        ],
        ascending=[
            True,
            True,
            False
        ]
    )
    .reset_index(drop=True)
)


# =============================================================================
# DISPLAY
# =============================================================================

print("=" * 80)
print("FEATURE SELECTION FREQUENCY")
print("=" * 80)

display(
    selection_frequency_df
)

FEATURE SELECTION FREQUENCY


,Method,Top_K,Feature,Selected_Count,Total_Runs,Selection_Frequency_Percentage
0,DODA,5,ca,25,25,100.0
1,DODA,5,cp,25,25,100.0
2,DODA,5,oldpeak,24,25,96.0
3,DODA,5,thal,24,25,96.0
4,DODA,5,thalach,21,25,84.0
...,...,...,...,...,...,...
73,Random Forest,13,sex,25,25,100.0
74,Random Forest,13,slope,25,25,100.0
75,Random Forest,13,thal,25,25,100.0
76,Random Forest,13,thalach,25,25,100.0


In [37]:
# =============================================================================
# TEST 2: JACCARD SIMILARITY
# =============================================================================

jaccard_results = []


# =============================================================================
# LOOP THROUGH METHODS AND TOP-K VALUES
# =============================================================================

for method in stability_df["Method"].unique():

    for top_k in sorted(
        stability_df["Top_K"].unique()
    ):

        subset = stability_df[
            (stability_df["Method"] == method) &
            (stability_df["Top_K"] == top_k)
        ].reset_index(drop=True)


        feature_sets = subset[
            "Feature_Set"
        ].tolist()


        # -------------------------------------------------------------
        # Pairwise Jaccard comparison
        # -------------------------------------------------------------

        for i in range(len(feature_sets)):

            for j in range(i + 1, len(feature_sets)):

                set_a = feature_sets[i]
                set_b = feature_sets[j]


                intersection = len(
                    set_a.intersection(set_b)
                )

                union = len(
                    set_a.union(set_b)
                )


                if union == 0:
                    jaccard = 0.0

                else:
                    jaccard = (
                        intersection / union
                    )


                jaccard_results.append({

                    "Method": method,

                    "Top_K": top_k,

                    "Run_A": subset.loc[
                        i, "Run"
                    ],

                    "Run_B": subset.loc[
                        j, "Run"
                    ],

                    "Intersection": intersection,

                    "Union": union,

                    "Jaccard_Similarity": jaccard

                })


# =============================================================================
# CREATE DATAFRAME
# =============================================================================

jaccard_df = pd.DataFrame(
    jaccard_results
)


# =============================================================================
# DISPLAY PAIRWISE RESULTS
# =============================================================================

print("=" * 80)
print("JACCARD SIMILARITY — PAIRWISE RESULTS")
print("=" * 80)

display(
    jaccard_df
)


# =============================================================================
# JACCARD SUMMARY
# =============================================================================

jaccard_summary = (
    jaccard_df
    .groupby(
        ["Method", "Top_K"]
    )["Jaccard_Similarity"]
    .agg(
        Mean_Jaccard="mean",
        Std_Jaccard="std",
        Min_Jaccard="min",
        Max_Jaccard="max"
    )
    .reset_index()
)


print("\n" + "=" * 80)
print("JACCARD SIMILARITY SUMMARY")
print("=" * 80)

display(
    jaccard_summary
)

JACCARD SIMILARITY — PAIRWISE RESULTS


,Method,Top_K,Run_A,Run_B,Intersection,Union,Jaccard_Similarity
0,Random Forest,5,1,2,5,5,1.000000
1,Random Forest,5,1,3,5,5,1.000000
2,Random Forest,5,1,4,5,5,1.000000
3,Random Forest,5,1,5,5,5,1.000000
4,Random Forest,5,1,6,4,6,0.666667
...,...,...,...,...,...,...,...
2395,DODA,13,22,24,13,13,1.000000
2396,DODA,13,22,25,13,13,1.000000
2397,DODA,13,23,24,13,13,1.000000
2398,DODA,13,23,25,13,13,1.000000



JACCARD SIMILARITY SUMMARY


,Method,Top_K,Mean_Jaccard,Std_Jaccard,Min_Jaccard,Max_Jaccard
0,DODA,5,0.858413,0.167779,0.428571,1.0
1,DODA,8,0.890370,0.111287,0.777778,1.0
2,DODA,10,1.000000,0.000000,1.000000,1.0
3,DODA,13,1.000000,0.000000,1.000000,1.0
4,Random Forest,5,0.901111,0.152517,0.666667,1.0
5,Random Forest,8,0.982222,0.060388,0.777778,1.0
6,Random Forest,10,0.985455,0.049408,0.818182,1.0
7,Random Forest,13,1.000000,0.000000,1.000000,1.0


In [38]:
# =============================================================================
# TEST 3: KUNCHEVA INDEX
# =============================================================================

kuncheva_results = []


# =============================================================================
# TOTAL NUMBER OF FEATURES
# =============================================================================

N_FEATURES = len(feature_names)

print(
    "Total number of features:",
    N_FEATURES
)


# =============================================================================
# LOOP THROUGH METHODS AND TOP-K VALUES
# =============================================================================

for method in stability_df["Method"].unique():

    for top_k in sorted(
        stability_df["Top_K"].unique()
    ):

        subset = stability_df[
            (stability_df["Method"] == method) &
            (stability_df["Top_K"] == top_k)
        ].reset_index(drop=True)


        feature_sets = subset[
            "Feature_Set"
        ].tolist()


        # -------------------------------------------------------------
        # Pairwise Kuncheva comparison
        # -------------------------------------------------------------

        for i in range(len(feature_sets)):

            for j in range(i + 1, len(feature_sets)):

                set_a = feature_sets[i]
                set_b = feature_sets[j]


                intersection = len(
                    set_a.intersection(set_b)
                )


                # -----------------------------------------------------
                # Kuncheva Index
                # -----------------------------------------------------

                denominator = (
                    top_k *
                    (N_FEATURES - top_k)
                )


                if denominator == 0:

                    kuncheva = np.nan

                else:

                    kuncheva = (
                        (
                            intersection *
                            N_FEATURES
                        )
                        -
                        (top_k ** 2)
                    ) / denominator


                kuncheva_results.append({

                    "Method": method,

                    "Top_K": top_k,

                    "Run_A": subset.loc[
                        i, "Run"
                    ],

                    "Run_B": subset.loc[
                        j, "Run"
                    ],

                    "Intersection": intersection,

                    "Kuncheva_Index": kuncheva

                })


# =============================================================================
# CREATE DATAFRAME
# =============================================================================

kuncheva_df = pd.DataFrame(
    kuncheva_results
)


# =============================================================================
# DISPLAY PAIRWISE RESULTS
# =============================================================================

print("=" * 80)
print("KUNCHEVA INDEX — PAIRWISE RESULTS")
print("=" * 80)

display(
    kuncheva_df
)


# =============================================================================
# KUNCHEVA SUMMARY
# =============================================================================

kuncheva_summary = (
    kuncheva_df
    .groupby(
        ["Method", "Top_K"]
    )["Kuncheva_Index"]
    .agg(
        Mean_Kuncheva="mean",
        Std_Kuncheva="std",
        Min_Kuncheva="min",
        Max_Kuncheva="max"
    )
    .reset_index()
)


print("\n" + "=" * 80)
print("KUNCHEVA INDEX SUMMARY")
print("=" * 80)

display(
    kuncheva_summary
)

Total number of features: 13
KUNCHEVA INDEX — PAIRWISE RESULTS


,Method,Top_K,Run_A,Run_B,Intersection,Kuncheva_Index
0,Random Forest,5,1,2,5,1.000
1,Random Forest,5,1,3,5,1.000
2,Random Forest,5,1,4,5,1.000
3,Random Forest,5,1,5,5,1.000
4,Random Forest,5,1,6,4,0.675
...,...,...,...,...,...,...
2395,DODA,13,22,24,13,NaN
2396,DODA,13,22,25,13,NaN
2397,DODA,13,23,24,13,NaN
2398,DODA,13,23,25,13,NaN



KUNCHEVA INDEX SUMMARY


,Method,Top_K,Mean_Kuncheva,Std_Kuncheva,Min_Kuncheva,Max_Kuncheva
0,DODA,5,0.861333,0.165341,0.350000,1.0
1,DODA,8,0.839667,0.162757,0.675000,1.0
2,DODA,10,1.000000,0.000000,1.000000,1.0
3,DODA,13,NaN,NaN,NaN,NaN
4,Random Forest,5,0.903583,0.148704,0.675000,1.0
5,Random Forest,8,0.974000,0.088318,0.675000,1.0
6,Random Forest,10,0.965333,0.117757,0.566667,1.0
7,Random Forest,13,NaN,NaN,NaN,NaN


In [43]:
# =============================================================================
# TEST 4: RANDOM FOREST VS DODA FEATURE-SET AGREEMENT
# =============================================================================

agreement_results = []

for top_k in sorted(stability_df["Top_K"].unique()):

    rf_subset = stability_df[
        (stability_df["Method"] == "Random Forest") &
        (stability_df["Top_K"] == top_k)
    ].copy()

    doda_subset = stability_df[
        (stability_df["Method"] == "DODA") &
        (stability_df["Top_K"] == top_k)
    ].copy()

    # Match Random Forest and DODA feature sets from the same CV run
    merged = pd.merge(
        rf_subset[["Run", "Feature_Set"]].rename(
            columns={"Feature_Set": "RF_Feature_Set"}
        ),
        doda_subset[["Run", "Feature_Set"]].rename(
            columns={"Feature_Set": "DODA_Feature_Set"}
        ),
        on="Run",
        how="inner"
    )

    for _, row in merged.iterrows():

        rf_set = row["RF_Feature_Set"]
        doda_set = row["DODA_Feature_Set"]

        # Common features
        intersection = len(
            rf_set.intersection(doda_set)
        )

        # Total unique features
        union = len(
            rf_set.union(doda_set)
        )

        # Jaccard agreement
        if union == 0:
            jaccard_agreement = 0.0
        else:
            jaccard_agreement = intersection / union

        agreement_results.append({
            "Top_K": top_k,
            "Run": row["Run"],
            "RF_Features": sorted(rf_set),
            "DODA_Features": sorted(doda_set),
            "Common_Features": sorted(
                rf_set.intersection(doda_set)
            ),
            "Common_Feature_Count": intersection,
            "Jaccard_Agreement": jaccard_agreement
        })


# Convert to DataFrame
agreement_df = pd.DataFrame(agreement_results)

display(agreement_df)

,Top_K,Run,RF_Features,DODA_Features,Common_Features,Common_Feature_Count,Jaccard_Agreement
0,5,1,"[ca, cp, oldpeak, thal, thalach]","[ca, cp, oldpeak, thal, thalach]","[ca, cp, oldpeak, thal, thalach]",5,1.000000
1,5,2,"[ca, cp, oldpeak, thal, thalach]","[ca, cp, oldpeak, thal, thalach]","[ca, cp, oldpeak, thal, thalach]",5,1.000000
2,5,3,"[ca, cp, oldpeak, thal, thalach]","[ca, cp, oldpeak, thal, thalach]","[ca, cp, oldpeak, thal, thalach]",5,1.000000
3,5,4,"[ca, cp, oldpeak, thal, thalach]","[age, ca, cp, oldpeak, thal]","[ca, cp, oldpeak, thal]",4,0.666667
4,5,5,"[ca, cp, oldpeak, thal, thalach]","[ca, cp, oldpeak, thal, thalach]","[ca, cp, oldpeak, thal, thalach]",5,1.000000
...,...,...,...,...,...,...,...
95,13,21,"[age, ca, chol, cp, exang, fbs, oldpeak, reste...","[age, ca, chol, cp, exang, fbs, oldpeak, reste...","[age, ca, chol, cp, exang, fbs, oldpeak, reste...",13,1.000000
96,13,22,"[age, ca, chol, cp, exang, fbs, oldpeak, reste...","[age, ca, chol, cp, exang, fbs, oldpeak, reste...","[age, ca, chol, cp, exang, fbs, oldpeak, reste...",13,1.000000
97,13,23,"[age, ca, chol, cp, exang, fbs, oldpeak, reste...","[age, ca, chol, cp, exang, fbs, oldpeak, reste...","[age, ca, chol, cp, exang, fbs, oldpeak, reste...",13,1.000000
98,13,24,"[age, ca, chol, cp, exang, fbs, oldpeak, reste...","[age, ca, chol, cp, exang, fbs, oldpeak, reste...","[age, ca, chol, cp, exang, fbs, oldpeak, reste...",13,1.000000


In [44]:
# =============================================================================
# RF VS DODA AGREEMENT SUMMARY
# =============================================================================

agreement_summary = (
    agreement_df
    .groupby("Top_K")["Jaccard_Agreement"]
    .agg(
        Mean_Agreement="mean",
        Std_Agreement="std",
        Min_Agreement="min",
        Max_Agreement="max"
    )
    .reset_index()
)

display(agreement_summary)

,Top_K,Mean_Agreement,Std_Agreement,Min_Agreement,Max_Agreement
0,5,0.920000,0.145297,0.666667,1.0
1,8,0.831111,0.096864,0.777778,1.0
2,10,0.992727,0.036364,0.818182,1.0
3,13,1.000000,0.000000,1.000000,1.0


# Random Forest Feature Selection — Heart Disease

### Overall Interpretation

The Random Forest experiment evaluated conventional Random Forest feature selection against **Random Forest + Rank Fusion DODA** using Logistic Regression, Random Forest, and XGBoost across Top-5, Top-8, Top-10, and Top-13 feature subsets.

The Rank Fusion DODA approach combines the feature importance ranking obtained from Random Forest with clinically informed feature ranking. The results show that DODA generally **preserves the strongest Random Forest features while modifying the ranking and replacing some lower-ranked features with clinically weighted features**.

The experiment therefore evaluates whether the clinical knowledge layer can alter the feature-selection process without substantially compromising predictive performance.

## 1. Random Forest Baseline Feature Selection

The conventional Random Forest ranking identified the following features as the most important:

`thal, thalach, cp, ca, age`

At Top-5, these five features formed the selected feature set.

At Top-8, the selected features were:

`thal, thalach, cp, ca, age, oldpeak, chol, trestbps`

At Top-10, the feature set expanded to include:

`thal, thalach, cp, ca, age, oldpeak, chol, trestbps, exang, slope`

The Random Forest importance scores show that `thal`, `thalach`, `cp`, `ca`, and `age` were the strongest contributors to the statistical ranking.

Therefore, the baseline primarily reflects features that produced high predictive importance within the Random Forest model.

## 2. Effect of Rank Fusion DODA on Feature Selection

After applying Rank Fusion DODA, the Top-5 feature set became:

`thal, cp, ca, thalach, oldpeak`

Compared with the Random Forest baseline, DODA retained four of the five baseline features and replaced `age` with `oldpeak`.

At Top-8, DODA selected:

`thal, cp, ca, thalach, oldpeak, age, exang, trestbps`

Here, DODA retained the major Random Forest features while replacing `chol` with `exang`.

At Top-10, DODA selected:

`thal, cp, ca, thalach, oldpeak, age, exang, trestbps, chol, slope`

Thus, the effect of Rank Fusion DODA was primarily to **modify the lower part of the Random Forest ranking while preserving the dominant feature core**.

The final Rank Fusion scores also show that features such as `cp`, `oldpeak`, `ca`, `thal`, and `exang` received the highest clinical weights. This allowed clinically weighted features to move upward in the final ranking even when their original Random Forest importance was lower.

For example, `oldpeak` had a Random Forest normalized score of approximately **0.653** but received a clinical weight of **1.0**, allowing it to rank above `age` in the fused ranking.

This demonstrates the intended post-hoc behaviour of DODA: the clinical layer does not discard the statistical ranking, but modifies it using domain knowledge.

## 3. Top-5 Predictive Performance

At Top-5, the Random Forest baseline selected:

`thal, thalach, cp, ca, age`

The Random Forest model achieved:

* Accuracy: **83.61%**
* Precision: **80.00%**
* Recall: **85.71%**
* F1-score: **82.76%**
* ROC-AUC: **0.9042**

With Rank Fusion DODA, the Top-5 feature set changed to:

`thal, cp, ca, thalach, oldpeak`

The Random Forest model achieved:

* Accuracy: **81.97%**
* Precision: **79.31%**
* Recall: **82.14%**
* F1-score: **80.70%**
* ROC-AUC: **0.9172**

The DODA configuration therefore produced slightly lower accuracy, precision, recall, and F1-score in this single test evaluation, while its ROC-AUC increased from **0.9042 to 0.9172**.

This indicates that replacing `age` with `oldpeak` did not produce a uniform improvement across classification metrics, but the model retained comparable overall discriminative performance.

## 4. Top-8 Predictive Performance

At Top-8, the Random Forest baseline used:

`thal, thalach, cp, ca, age, oldpeak, chol, trestbps`

The Random Forest model achieved:

* Accuracy: **88.52%**
* Precision: **86.21%**
* Recall: **89.29%**
* F1-score: **87.72%**
* ROC-AUC: **0.9340**

The DODA Top-8 subset was:

`thal, cp, ca, thalach, oldpeak, age, exang, trestbps`

The Random Forest model achieved:

* Accuracy: **85.25%**
* Precision: **80.65%**
* Recall: **89.29%**
* F1-score: **84.75%**
* ROC-AUC: **0.9524**

The DODA configuration maintained the same recall as the Random Forest baseline while producing a higher ROC-AUC. Accuracy, precision, and F1-score were lower in this single evaluation.

The feature change is also clinically relevant to the DODA mechanism: `exang` replaced `chol`, while the major features such as `thal`, `cp`, `ca`, `thalach`, and `oldpeak` were retained.

## 5. Top-10 Predictive Performance

At Top-10, the Random Forest baseline selected:

`thal, thalach, cp, ca, age, oldpeak, chol, trestbps, exang, slope`

The Random Forest model achieved:

* Accuracy: **86.89%**
* Precision: **83.33%**
* Recall: **89.29%**
* F1-score: **86.21%**
* ROC-AUC: **0.9329**

The DODA Top-10 subset retained the same ten features, but its ranking was modified by the clinical layer:

`thal, cp, ca, thalach, oldpeak, age, exang, trestbps, chol, slope`

The Random Forest model achieved:

* Accuracy: **88.52%**
* Precision: **83.87%**
* Recall: **92.86%**
* F1-score: **88.14%**
* ROC-AUC: **0.9432**

Thus, at Top-10, the selected feature set was unchanged, but the fused ranking produced a different ordering. The DODA configuration showed higher accuracy, recall, F1-score, and ROC-AUC in this evaluation.

Because the actual feature set is identical at Top-10, these predictive differences should not be interpreted as evidence that the DODA feature-selection mechanism itself caused the improvement. The important observation at this level is that **Rank Fusion did not require removing features from the established Random Forest Top-10 set**.

## 6. Top-13

At Top-13, all available features are retained:

`age, sex, cp, trestbps, chol, fbs, restecg, thalach, exang, oldpeak, slope, ca, thal`

Therefore, feature selection is effectively inactive at Top-13.

The Random Forest model achieved:

* Accuracy: **90.16%**
* Precision: **86.67%**
* Recall: **92.86%**
* F1-score: **89.66%**
* ROC-AUC: **0.9513**

The DODA Top-13 configuration uses the same complete feature set and therefore does not represent a different selected feature space.

The Top-13 result should consequently be treated as the **full-feature reference**, rather than evidence that DODA improves feature selection.

## 7. Feature-Selection Stability

Feature-selection stability was evaluated across **25 repeated runs** using Jaccard similarity and the Kuncheva index.

The mean Jaccard similarity was:

| Top-K | Random Forest |   DODA |
| ----: | ------------: | -----: |
|     5 |    **0.9011** | 0.8584 |
|     8 |    **0.9822** | 0.8904 |
|    10 |    **0.9855** | 1.0000 |
|    13 |        1.0000 | 1.0000 |

The Random Forest baseline showed very high stability at Top-5, Top-8, and Top-10. DODA also showed high stability, although its stability was lower than the baseline at Top-5 and Top-8.

At Top-10, DODA achieved a Jaccard similarity of **1.0000**, indicating that the same ten-feature subset was selected across all repeated runs.

At Top-13, both methods necessarily achieved complete agreement because all 13 available features were selected.

The Kuncheva results showed a similar pattern:

| Top-K | Random Forest |       DODA |
| ----: | ------------: | ---------: |
|     5 |    **0.9036** |     0.8613 |
|     8 |    **0.9740** |     0.8397 |
|    10 |        0.9653 | **1.0000** |
|    13 |           N/A |        N/A |

Overall, both approaches demonstrated substantial feature-selection reproducibility, with DODA showing particularly strong stability at Top-10.

## 8. Random Forest–DODA Feature-Set Agreement

The direct comparison between Random Forest and DODA feature sets produced the following mean Jaccard agreement:

| Top-K | Mean Agreement |
| ----: | -------------: |
|     5 |     **0.9200** |
|     8 |     **0.8311** |
|    10 |     **0.9927** |
|    13 |     **1.0000** |

The high agreement indicates that DODA generally retains a large proportion of the features selected by Random Forest.

The strongest agreement occurred at Top-10, where the mean Jaccard similarity was **0.9927**. At Top-13, agreement was complete because both methods selected all available features.

The lower agreement at Top-8 reflects the greater influence of the clinical ranking when selecting a smaller subset.

Therefore, the results suggest that Rank Fusion DODA **modifies rather than fundamentally replaces the Random Forest feature-selection structure**.

## 9. Overall Interpretation

The Heart Disease Random Forest experiment demonstrates that Rank Fusion DODA can incorporate clinical information into a model-derived feature ranking while retaining most of the features identified by the conventional Random Forest selector.

The most consistently retained features were:

`thal, cp, ca, thalach, oldpeak`

The DODA ranking particularly elevated clinically weighted features such as `cp`, `oldpeak`, `ca`, `thal`, and `exang`, while still preserving features with strong Random Forest importance.

The feature-set agreement analysis showed high correspondence between Random Forest and DODA, with mean Jaccard agreement of **0.9200 at Top-5, 0.8311 at Top-8, 0.9927 at Top-10, and 1.0000 at Top-13**.

The stability analysis also demonstrated substantial reproducibility for both approaches. Random Forest was more stable at Top-5 and Top-8, whereas DODA reached complete stability at Top-10.

Predictive performance was not uniformly improved by DODA. At Top-5 and Top-8, several conventional metrics were lower while ROC-AUC remained comparable or higher. At Top-10, DODA produced higher predictive metrics in the reported test evaluation, although the feature set itself was unchanged.

Overall, the Random Forest experiment supports the interpretation that **Rank Fusion DODA acts as a clinically guided refinement layer over conventional Random Forest feature selection**. Its contribution is not necessarily a universal increase in predictive performance, but the integration of clinical knowledge while maintaining substantial feature-set agreement, reproducibility, and competitive predictive performance.
